### DISCUSSION TITLE
Option for early stopping during training

### DISCUSSION THREAD

**Discussion by LennardLux** [neuralhydrology/neuralhydrology#139] [Q&A] [answered]
Hi, I'm currently training multiple models simultaneously, all for the same number of epochs. However, I've observed that some of the models seem to reach their optimum earlier than others during training. Is there a way to implement early stopping for these models to save training time? I'd appreciate any insights or ideas!

**Comment by gauchm** [comment, marked as answer]
We currently don't have an option for early stopping. But (at least a basic version) should be quite easy to implement in [basetrainer.py](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/training/basetrainer.py):
Check the validation metrics in [L230](https://github.com/neuralhydrology/neuralhydrology/blob/1a3c313a83832796c5d7a849ca02dc09f7dbb10a/neuralhydrology/training/basetrainer.py#L230) and keep track of the number of epochs without improvement. After a specified maximum patience, `break` out of the for-loop over epochs.

For a quick-and-dirty implementation, you could just hard-code your desired patience & validation metric (either the validation loss or one of the metrics that are actually calculated during validation as specified in the config.yml). A nice implementation would make these choices configurable via [config.py](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/utils/config.py).

**Reply by LennardLux** [reply]
Thanks for the quick response, I will try that :)
